src/os/bluestore/BlueFS.cc

```diff
@@ -3038,54 +3036,67 @@ void BlueFS::_consume_dirty(uint64_t seq)
-     * put update op(log) to log and we are fine.
-     * Problem - it interferes with log compaction:
-     * New log produced in compaction will include - as last op - jump into some offset (anchor) of current log.
-     * It is assumed that log region (anchor - end) will contain all changes made by bluefs since
-     * full state capture into new log.
-     * Putting log update into (anchor - end) region is illegal, because any update there must be compatible with
-     * both logs, but old log is different then new log.
-     *
-     * Possible solutions:
-     * - stall extending log until we finish compacting and switch log (CURRENT)
-     * - re-run compaction with more runway for old log
-     * - add OP_FILE_ADDEXT that adds extent; will be compatible with both logs
-     */
-    if (log_forbidden_to_expand.load() == true) {
-      return -EWOULDBLOCK;
+  if (!ignore_compaction) {
+    std::unique_lock<ceph::mutex> ll(log.lock, std::adopt_lock);
+    while (log_forbidden_to_expand.load() == true) {
+      log_cond.wait(ll);
     }
-    vselector->sub_usage(log.writer->file->vselector_hint, log.writer->file->fnode);
+    ll.release();
+  }
+  vselector->sub_usage(log.writer->file->vselector_hint, log.writer->file->fnode);
+  while (amount > 0) {
     int r = _allocate(
-      vselector->select_prefer_bdev(log.writer->file->vselector_hint),
-      cct->_conf->bluefs_max_log_runway,
-      0,
-      &log.writer->file->fnode);
+        vselector->select_prefer_bdev(log.writer->file->vselector_hint),
+        cct->_conf->bluefs_max_log_runway,
+        0,
+        &log.writer->file->fnode);
+    amount -= std::min(cct->_conf->bluefs_max_log_runway, amount);
     ceph_assert(r == 0);
-    vselector->add_usage(log.writer->file->vselector_hint, log.writer->file->fnode);
-    log.t.op_file_update_inc(log.writer->file->fnode);
   }
-  return runway;
+  vselector->add_usage(log.writer->file->vselector_hint, log.writer->file->fnode);
+
+  bluefs_transaction_t log_extend_transaction;
+  log_extend_transaction.seq = log.t.seq;
+  log_extend_transaction.uuid = log.t.uuid;
+  log_extend_transaction.op_file_update_inc(log.writer->file->fnode);
+  bufferlist bl;
+  bl.reserve(super.block_size);
+  encode(log_extend_transaction, bl);
+  size_t realign = super.block_size - (bl.length() % super.block_size);
+  if (realign && realign != super.block_size)
+    bl.append_zero(realign);
+  log.writer->append(bl);
+  uint64_t new_data = _flush_special(log.writer);
+  vselector->add_usage(log.writer->file->vselector_hint, new_data);
+
+  return log.writer->file->fnode.get_allocated() - log.writer->get_effective_write_pos();
 }
 
-void BlueFS::_flush_and_sync_log_core(int64_t runway)
+void BlueFS::_flush_and_sync_log_core(bool ignore_compaction)
 {
   ceph_assert(ceph_mutex_is_locked(log.lock));
   dout(10) << __func__ << " " << log.t << dendl;
 
+  _maybe_extend_log(log.t.expected_size(), ignore_compaction);
```

Comment:
But without this extension of log it would be possible for a larger transaction to be written and then, if it was larger, it will be overwritten because you jump back to the tail extent:

transaction:
```
                      end of current extents        end of extents with tail
                          v                          v
before tail -> `AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA`
after tail  -> `AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA`
```
In that case we still need to increse before adding the tail to ensure transaction has enough space because after flushing that transaction it will jump back to _end of extents with tail_

Reply:
We need to protect primarily against log writes coming from `_flush_and_sync_log`. This is the most frequently used path.
But path coming from `_compact_log_async` cannot be ignored, and async compaction the way we currently do (injection of new starting point compatible with existing starting point) is a very delicate process.